### PR TITLE
[Serializer] [PropertyAccessor] Ignore non-collection interface generics

### LINKED ISSUE #41996: [Serializer] Allow using generic annotations and not parsing them as an array
**Symfony version(s) affected**: 5.3.3

**Description**
Denormalizer breaks with an error `Data expected to be an array, string given.` when it handles an object, which contains  a generic-like annotation, but is not an array or iterable.

**How to reproduce**
```php
interface Foo {}

/** @template TFoo of Foo */
interface Bar {}

final class FooBar {
    /** @var Bar<Foo> */
    public Bar $bar; // This prop will fail
    // ...
}
```

**Possible Solution**
I digged it in PhpDocTypeHelper, which is part of phpDocumentor, so I don't know how to solve it. :worried:

### PR DESCRIPTION
**Pull request by mtarld**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #41996 #49924
| License       | MIT

`PhpDocExtractor` and `PhpStanDocExtractor` are following an old version of PSR-5 with used to define collections as the following:
```
generic = collection-type "<" [type-expression "," *SP] type-expression ">"
```
But, it does conflict with non-collection generics.

This issue will automatically be solved if the `TypeInfo` is merged in Symfony. But for older versions (<7.1 at least), it needs a fix.

~I was not able to find a proper bug fix without introducing a BC break, so this PR proposes to opt-on the "non-collection generics" parsing, such as `stcClass<int>` for example.~

~To opt-out from parsing these generics, it'll be required to set `ignore_docblock_generics` in the context. And this key/value will become obsolete as soon as the `TypeInfo` is introduced.~

~I'm not sure how to proceed though, should we only merge it in 5.4, and 6.3 supposing that the `TypeInfo` might be merged in 7.x? Should we document it only in these branches?~

EDIT: I finally ignored PHPDoc generics that aren't well known collection generic types so that the process will fall back to other resolvers (such as reflection resolver for example)

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
What about considering this as an improvement and merging into 6.4? Then we can do the BC break in 7.0.
I'd be fine saying those annotations aren't supported in 5.4 personnally.

**Comment by mtarld** [comment]
It also means that it won't be supported in 6.3 and 7.0 (supposing that the `TypeInfo` will come in 7.1). But I'm fine with that as well, as 6.4 matters the most, being an LTS.

**Review by maxhelias — APPROVED** [review]

File: src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php [hunk 1]
```diff
@@ -115,10 +115,17 @@ private function createType(DocType $type, bool $nullable, ?string $docType = nu
 
             [$phpType, $class] = $this->getPhpTypeAndClass((string) $fqsen);
 
+            $collection = \in_array($class, [\Traversable::class, \Iterator::class, \IteratorAggregate::class, \ArrayAccess::class, \Generator::class], true);
```
Comment (stof) [inline src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: shouldn't we at least support the Collection and ReadableCollection interfaces of doctrine/collections there, to limit the impact of the BC break ?

Comment (mtarld) [reply src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: I think that by doing the following: `$collection = \is_a($class, \Traversable::class, true) || \is_a($class, \ArrayAccess::class, true);`, most of the collections will be handled properly.

Comment (mtarld) [reply src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: Similar is being done in https://github.com/symfony/symfony/pull/54661

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @mtarld.

**Comment by xabbuh** [comment]
@mtarld I have merged these changes up to the `7.1` branch which now fails with one error. I tried to find out what we need to change but wasn't able to resolve it. In case you want to take a look at it. :)

**Comment by mtarld** [comment]
@xabbuh, indeed new tests must be updated like the current ones. I created PR fixing it: https://github.com/symfony/symfony/pull/57530.
Thanks for reaching me :slightly_smiling_face:

**Comment by soyuka** [comment]
HI, since that we're observing a `collection: false` instead of `true` for properties typed as: 

```
\Doctrine\Common\Collections\ArrayCollection<Foo>
```

`is_a(ArrayCollection::class, \ArrayAccess::class, true)` is `true` not sure why it fails, any idea @mtarld ?

**Comment by andyexeter** [comment]
We're seeing the same issue as @soyuka. Looks like it will be fixed once #57617 is merged

**Pull request merged**